Added requested_bytes to CUDA Caching Allocator Stats

c-odrin opened the pull request:
Summary:
The caching allocator can be configured to round memory allocations in order to reduce fragmentation. Sometimes however, the overhead from rounding can be higher than the fragmentation it helps reduce.

We have added a new stat to CUDA caching allocator stats to help track if rounding is adding too much overhead and help tune the roundup_power2_divisions flag:
    - "requested_bytes.{current,peak,allocated,freed}": memory requested by client code, compare this with allocated_bytes to check if allocation rounding adds too much overhead

Test Plan: Added test case in caffe2/test/test_cuda.py

Differential Revision: D40810674

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -245,12 +249,14 @@ struct AllocParams {
   AllocParams(
       int device,
       size_t size,
+      size_t requested_size,
```
Might be simpler to just initialized requested_size when the block is set to allocated rather than when a block is created.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -198,12 +200,14 @@ struct Block {
         stream(stream),
         stream_uses(),
         size(size),
+        requested_size(requested_size),
```
We can initialize this to 0 and set it to to the requested size when the block becomes allocated.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -954,7 +962,7 @@ class DeviceCachingAllocator {
       remaining = block;
 
-      block = new Block(device, stream, size, &pool, block->ptr);
+      block = new Block(device, stream, size, orig_size, &pool, block->ptr);
```
Blocks can be used multiple times (in particular the not split case below). This only assigns the requested size to the block on the first use but it won't get updated in the other case. Maybe try setting the requested_bytes at the same time when the block is marked allocated since every new use of a block has to be coupled with the block being marked allocated.

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -90,6 +90,9 @@ struct DeviceStats {
   // COUNT: total number of oversize blocks requiring malloc
   Stat oversize_segments;
 
+  // SUM: bytes requested by client code
```
Is there a reason this wouldn't mirror `allocated_bytes` in recording min/max etc?

c-odrin replied:
I've now implemented this like the rest of the stats e.g. `allocated_bytes`, `reserved_bytes`

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -121,6 +125,7 @@ struct SegmentInfo {
   int64_t device = 0;
   int64_t address = 0;
   int64_t total_size = 0;
+  int64_t total_requested_size = 0;
```
I can't find the code to exported requested_size/total_requested size.

zdevito reviewed (commented):
It makes sense to record this stat. I have a few inline comments. I also think that there is code missing to handle resetting the statistic when all the other statistics are reset. I don't see tests for reading requested_bytes out of the block info.

c-odrin commented:
> It makes sense to record this stat. I have a few inline comments. I also think that there is code missing to handle resetting the statistic when all the other statistics are reset. I don't see tests for reading requested_bytes out of the block info.

Thank you for the review! I've addressed the inline comments, added tests for reading requested_bytes out of segment info/block info and added code to reset the statistics when all the other statistics are reset.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -1435,7 +1460,12 @@ class DeviceCachingAllocator {
-      update_stat(stats.active_bytes[stat_type], -original_block_size);
+      update_stat(
+          stats.active_bytes[stat_type],
+          static_cast<std::int64_t>(-original_block_size));
```
Should these negations of unsigned quantities happen after the conversion to int64_t.

c-odrin replied:
Addressed

zdevito commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -633,6 +635,7 @@ PyObject* THCPModule_memorySnapshot(PyObject* _unused, PyObject* noargs) {
     segmentDict[total_size_s] = segmentInfo.total_size;
     segmentDict[allocated_size_s] = segmentInfo.allocated_size;
     segmentDict[active_size_s] = segmentInfo.active_size;
+    segmentDict[requested_size_s] = segmentInfo.requested_size;
```
Exporting BlockInfo `requested_size` is missing here and in `memory_snapshot.cpp` I also couldn't find a test for the blockinfo/segmentinfo information.

c-odrin replied:
Addressed - exporting BlockInfo `requested_size` and added test as part of _check_memory_stat_consistency

zdevito reviewed (approved):
Looks good. I have a couple nits that should be addressed (missing blockinfo export and test), but this otherwise looks good to me.

c-odrin committed:
Added requested_bytes to CUDA Caching Allocator Stats (#88575)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88575

The caching allocator can be configured to round memory allocations in order to reduce fragmentation. Sometimes however, the overhead from rounding can be higher than the fragmentation it helps reduce.

We have added a new stat to CUDA caching allocator stats to help track if rounding is adding too much overhead and help tune the roundup_power2_divisions flag:
    - "requested_bytes.{all,large_pool,small_pool}.{current,peak,allocated,freed}": memory requested by client code, compare this with allocated_bytes to check if allocation rounding adds too much overhead

Test Plan: Added test case in caffe2/test/test_cuda.py

Differential Revision: D40810674

fbshipit-source-id: d71624c0173c1ca272a1a76acc2bc1ff022edfab

c-odrin committed:
Merge branch 'master' into export-D40810674

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR.  This change seems to cause libtorch build to run out of memory on a 4xlarge runner https://hud.pytorch.org/pytorch/pytorch/commit/54b7c7d5e905b43584aa197006961e6b395ce574.  Docker crashes with exit code 137 clearly means running OOM' -c weird
